File: oxidize-golang/core/model/speculative.go
package model

import (
	"fmt"
	"math"
)

// SpeculativeConfig mirrors SpeculativeConfig.
type SpeculativeConfig struct {
	DraftTokensPerStep int
	MaxNewTokens       int
	Sampling           SamplingConfig
	StopToken          *Token
	StrictMode         bool
	MinAcceptanceRate  float32
}

// DefaultSpeculativeConfig returns the balanced default.
func DefaultSpeculativeConfig() SpeculativeConfig {
	return SpeculativeConfig{
		DraftTokensPerStep: 4,
		MaxNewTokens:       128,
		Sampling:           DefaultSamplingConfig(),
		StopToken:          nil,
		StrictMode:         false,
		MinAcceptanceRate:  0.3,
	}
}

// Conservative returns a config tuned for high accuracy (low draft count).
func Conservative() SpeculativeConfig {
	c := DefaultSpeculativeConfig()
	c.DraftTokensPerStep = 2
	c.StrictMode = true
	return c
}

// Aggressive returns a config tuned for speed (high draft count, lower
// acceptance threshold).
func Aggressive() SpeculativeConfig {
	c := DefaultSpeculativeConfig()
	c.DraftTokensPerStep = 8
	c.MinAcceptanceRate = 0.3
	return c
}

// SpeculativeStats mirrors SpeculativeStats.
type SpeculativeStats struct {
	TotalDraftTokens     int
	TotalAcceptedTokens  int
	TotalRejectedTokens  int
	DraftForwardPasses   int
	TargetForwardPasses  int
	FallbackTokens       int
	Accepted             int
	Rejected             int
	Total                int
}

// AcceptanceRate returns the ratio of accepted drafts to total.
func (s SpeculativeStats) AcceptanceRate() float32 {
	if s.Total == 0 {
		return 0
	}
	return float32(s.Accepted) / float32(s.Total)
}

// SpeculativeDecoder mirrors SpeculativeDecoder<'a, T: Model>.
type SpeculativeDecoder struct {
	Draft     Model
	Target    Model
	Session   *Session
	Config    SpeculativeConfig
	Stats     SpeculativeStats
}

// NewSpeculativeDecoder constructs a decoder with the given draft + target
// models.
func NewSpeculativeDecoder(draft, target Model, session *Session, config SpeculativeConfig) *SpeculativeDecoder {
	return &SpeculativeDecoder{Draft: draft, Target: target, Session: session, Config: config}
}

// Step runs one speculative decode step and returns the accepted tokens.
func (d *SpeculativeDecoder) Step() ([]Token, error) {
	if d.Config.DraftTokensPerStep <= 0 {
		return nil, fmt.Errorf("draft tokens per step must be > 0")
	}
	tokens, accepted, err := SpeculativeDecode(d.Draft, d.Target, d.Config.DraftTokensPerStep, d.Session)
	if err != nil {
		return nil, err
	}
	d.Stats.Total += d.Config.DraftTokensPerStep
	if accepted < d.Config.DraftTokensPerStep {
		d.Stats.Rejected++
	} else {
		d.Stats.Accepted += d.Config.DraftTokensPerStep
	}
	return tokens, nil
}

// SpeculativeError mirrors SpeculativeError.
type SpeculativeError struct{ Message string }

func (e *SpeculativeError) Error() string { return "speculative: " + e.Message }

// LoadDraftModelForSpeculative is a placeholder helper. In the Rust crate it
// resolves a smaller draft model from the same path; in the Go port we
// return a stub model with the same vocab/context as the target.
func LoadDraftModelForSpeculative(target Model) Model {
	switch m := target.(type) {
	case *LlamaModel:
		out := *m
		out.Config.LayerCount = m.Config.LayerCount / 2
		if out.Config.LayerCount == 0 {
			out.Config.LayerCount = 1
		}
		return &out
	case *InferenceModel:
		out := *m
		out.Config.LayerCount = m.Config.LayerCount / 2
		if out.Config.LayerCount == 0 {
			out.Config.LayerCount = 1
		}
		return &out
	}
	return target
}

// SpeculativeConfigBuilder mirrors SpeculativeConfigBuilder.
type SpeculativeConfigBuilder struct {
	cfg SpeculativeConfig
}

// NewSpeculativeConfigBuilder constructs a builder.
func NewSpeculativeConfigBuilder() *SpeculativeConfigBuilder { return &SpeculativeConfigBuilder{cfg: DefaultSpeculativeConfig()} }

// WithDraftTokens sets the draft tokens per step.
func (b *SpeculativeConfigBuilder) WithDraftTokens(n int) *SpeculativeConfigBuilder { b.cfg.DraftTokensPerStep = n; return b }

// WithMaxNewTokens sets the maximum new tokens.
func (b *SpeculativeConfigBuilder) WithMaxNewTokens(n int) *SpeculativeConfigBuilder { b.cfg.MaxNewTokens = n; return b }

// WithStrict sets strict mode.
func (b *SpeculativeConfigBuilder) WithStrict(s bool) *SpeculativeConfigBuilder { b.cfg.StrictMode = s; return b }

// Build returns the final config.
func (b *SpeculativeConfigBuilder) Build() SpeculativeConfig { return b.cfg }

// silence unused import warnings
var _ = math.Abs

File: oxidize-python/oxidize_python/core/model/speculative.py
"""Speculative decoding mirroring oxidize-golang/core/model/speculative.go."""

from __future__ import annotations

from dataclasses import dataclass

from oxidize_python.core.model.inference import InferenceModel
from oxidize_python.core.model.llama import LlamaModel
from oxidize_python.core.model.model import Model, Session, Token
from oxidize_python.core.model.sampling import (
    SamplingConfig,
    default_sampling_config,
    speculative_decode,
)


@dataclass
class SpeculativeConfig:
    draft_tokens_per_step: int = 4
    max_new_tokens: int = 128
    sampling: SamplingConfig | None = None
    stop_token: Token | None = None
    strict_mode: bool = False
    min_acceptance_rate: float = 0.3

    def __post_init__(self) -> None:
        if self.sampling is None:
            self.sampling = default_sampling_config()


def default_speculative_config() -> SpeculativeConfig:
    return SpeculativeConfig()


def conservative_speculative_config() -> SpeculativeConfig:
    c = default_speculative_config()
    c.draft_tokens_per_step = 2
    c.strict_mode = True
    return c


def aggressive_speculative_config() -> SpeculativeConfig:
    c = default_speculative_config()
    c.draft_tokens_per_step = 8
    c.min_acceptance_rate = 0.3
    return c


@dataclass
class SpeculativeStats:
    total_draft_tokens: int = 0
    total_accepted_tokens: int = 0
    total_rejected_tokens: int = 0
    draft_forward_passes: int = 0
    target_forward_passes: int = 0
    fallback_tokens: int = 0
    accepted: int = 0
    rejected: int = 0
    total: int = 0

    def acceptance_rate(self) -> float:
        if self.total == 0:
            return 0.0
        return self.accepted / self.total

    def record_step(
        self,
        draft_tokens: int,
        accepted_draft_tokens: int,
        used_residual_fallback: bool,
    ) -> None:
        """Record the outcome of a single speculative verification step.
        Mirrors SpeculativeStats.RecordStep in speculative.go."""
        self.total_draft_tokens += draft_tokens
        self.total_accepted_tokens += accepted_draft_tokens
        rejected = draft_tokens - accepted_draft_tokens
        if rejected < 0:
            rejected = 0
        self.total_rejected_tokens += rejected
        self.draft_forward_passes += 1
        self.target_forward_passes += 1
        if used_residual_fallback:
            self.fallback_tokens += 1
        self.total += draft_tokens
        self.accepted += accepted_draft_tokens
        if accepted_draft_tokens < draft_tokens:
            self.rejected += 1

    def draft_acceptance_rate(self) -> float:
        """Fraction of proposed draft tokens that were accepted. Mirrors
        SpeculativeStats.DraftAcceptanceRate in speculative.go."""
        if self.total_draft_tokens == 0:
            return 0.0
        return self.total_accepted_tokens / self.total_draft_tokens

    def tokens_per_target_forward(self) -> float:
        """Average number of emitted tokens produced per target forward pass.
        Mirrors SpeculativeStats.TokensPerTargetForward in speculative.go."""
        if self.target_forward_passes == 0:
            return 0.0
        emitted = self.total_accepted_tokens + self.fallback_tokens
        return emitted / self.target_forward_passes


class SpeculativeError(Exception):
    def __init__(self, message: str) -> None:
        super().__init__(f"speculative: {message}")


class SpeculativeDecoder:
    def __init__(
        self,
        draft: Model,
        target: Model,
        session: Session,
        config: SpeculativeConfig,
    ) -> None:
        self.draft = draft
        self.target = target
        self.session = session
        self.config = config
        self.stats = SpeculativeStats()

    def step(self) -> list[Token]:
        if self.config.draft_tokens_per_step <= 0:
            raise SpeculativeError("draft tokens per step must be > 0")
        tokens, accepted = speculative_decode(
            self.draft,
            self.target,
            self.config.draft_tokens_per_step,
            self.session,
        )
        self.stats.total += self.config.draft_tokens_per_step
        if accepted < self.config.draft_tokens_per_step:
            self.stats.rejected += 1
        else:
            self.stats.accepted += self.config.draft_tokens_per_step
        return tokens


def load_draft_model_for_speculative(target: Model) -> Model:
    if isinstance(target, LlamaModel):
        cfg = target.config
        return LlamaModel(
            architecture=cfg.architecture,
            vocab_size=cfg.vocab_size,
            context_size=cfg.context_size,
            layer_count=max(1, cfg.layer_count // 2),
        )
    if isinstance(target, InferenceModel):
        from dataclasses import replace

        cfg = replace(target.config, layer_count=max(1, target.config.layer_count // 2))
        return InferenceModel(config=cfg, storage=target.storage, stack=target.stack)
    return target


@dataclass
class SpeculativeConfigBuilder:
    cfg: SpeculativeConfig | None = None

    def __post_init__(self) -> None:
        if self.cfg is None:
            self.cfg = default_speculative_config()

    def with_draft_tokens(self, n: int) -> SpeculativeConfigBuilder:
        assert self.cfg is not None
        self.cfg.draft_tokens_per_step = n
        return self

    def with_max_new_tokens(self, n: int) -> SpeculativeConfigBuilder:
        assert self.cfg is not None
        self.cfg.max_new_tokens = n
        return self

    def with_strict(self, s: bool) -> SpeculativeConfigBuilder:
        assert self.cfg is not None
        self.cfg.strict_mode = s
        return self

    def build(self) -> SpeculativeConfig:
        assert self.cfg is not None
        return self.cfg
